clients/js/src/plugins/lifecycleHook.ts

```diff
@@ -0,0 +1,128 @@
+import { PublicKey } from '@metaplex-foundation/umi';
+import {
+  ExtraAccount,
+  extraAccountFromBase,
+  extraAccountToBase,
+} from './extraAccount';
+import {
+  BaseLifecycleHook,
+  BaseLifecycleHookInitInfoArgs,
+  BaseLifecycleHookUpdateInfoArgs,
+  ExternalPluginSchema,
+} from '../generated';
+import { LifecycleChecks, lifecycleChecksToBase } from './lifecycleChecks';
+import {
+  PluginAuthority,
+  pluginAuthorityFromBase,
+  pluginAuthorityToBase,
+} from './pluginAuthority';
+import { BaseExternalPlugin, parseExternalPluginData } from './externalPlugins';
+import { ExternalPluginManifest } from './externalPluginManifest';
+import { ExternalPluginKey } from './externalPluginKey';
+
+export type LifecycleHook = Omit<
+  BaseLifecycleHook,
+  'extraAccounts' | 'dataAuthority'
+> & {
+  extraAccounts?: Array<ExtraAccount>;
+  dataAuthority?: PluginAuthority;
+  data?: any;
+};
+
+export type LifecycleHookPlugin = BaseExternalPlugin &
+  LifecycleHook & {
+    type: 'LifecycleHook';
+    hookedProgram: PublicKey;
+  };
+
+export type LifecycleHookInitInfoArgs = Omit<
+  BaseLifecycleHookInitInfoArgs,
+  'initPluginAuthority' | 'lifecycleChecks' | 'schema' | 'dataAuthority'
+> & {
+  type: 'LifecycleHook';
+  initPluginAuthority?: PluginAuthority;
+  lifecycleChecks?: LifecycleChecks;
+  schema?: ExternalPluginSchema;
+  extraAccounts?: Array<ExtraAccount>;
+  dataAuthority?: PluginAuthority;
+};
+
+export type LifecycleHookUpdateInfoArgs = Omit<
+  BaseLifecycleHookUpdateInfoArgs,
+  'lifecycleChecks' | 'extraAccounts' | 'schema'
+> & {
+  key: ExternalPluginKey;
+  lifecycleChecks?: LifecycleChecks;
+  extraAccounts?: Array<ExtraAccount>;
+  schema?: ExternalPluginSchema;
+};
+
+export function lifecycleHookInitInfoArgsToBase(
+  l: LifecycleHookInitInfoArgs
+): BaseLifecycleHookInitInfoArgs {
+  return {
+    extraAccounts: l.extraAccounts
+      ? l.extraAccounts.map(extraAccountToBase)
+      : null,
+    hookedProgram: l.hookedProgram,
+    initPluginAuthority: l.initPluginAuthority
+      ? pluginAuthorityToBase(l.initPluginAuthority)
+      : null,
+    lifecycleChecks: l.lifecycleChecks
+      ? lifecycleChecksToBase(l.lifecycleChecks)
+      : null,
+    schema: l.schema ? l.schema : null,
+    dataAuthority: l.dataAuthority
+      ? pluginAuthorityToBase(l.dataAuthority)
+      : null,
+  };
+}
```

Comment:
I think it is a safe assumption to say overall that the main reason we create new types `Xyz` and rename the generated ones as `BaseXyz` is so that we have versions that are nicer to work with in general.

Probably the two main ones are :
1. So that we can have types where we don't provide some fields and those values are set to `null` in the base type.
2. And maybe also for dealing with Rust enums like in the case of `Authority` where we gave it a type and an address field.

Does that capture it pretty well or are there other reasons I should be seeing?

Reply:
correct, I explored ways to modify or add to the the generated types in kinobi but it was just quicker to hand roll it for now